trieste/models/gpflow/models.py

```diff
@@ -871,3 +895,181 @@ def get_internal_data(self) -> Dataset:
+        ]
+    )
+    return cov
+
+
+def compute_kernel_blocks(
+    kernel: gpflow.kernels.Kernel,
+    inducing_variable: InducingVariables,
+    query_points_1: TensorType,
+    query_points_2: TensorType,
+    num_latent: int,
+) -> tuple[TensorType, TensorType, TensorType, TensorType]:
+
+    if type(inducing_variable) in [gpflow.inducing_variables.SharedIndependentInducingVariables]:
+        inducing_points = inducing_variable.inducing_variable.Z
+    elif type(inducing_variable) in [
+        gpflow.inducing_variables.SeparateIndependentInducingVariables
+    ]:
+        inducing_points = [
+            inducing_variable.Z for inducing_variable in inducing_variable.inducing_variables
+        ]
+    else:
+        inducing_points = inducing_variable.Z
+
+    if type(kernel) in [gpflow.kernels.SharedIndependent, gpflow.kernels.SeparateIndependent]:
+        if type(inducing_points) == list:
+
+            K = tf.concat(
+                [ker(Z)[None, ...] for ker, Z in zip(kernel.kernels, inducing_points)], axis=0
+            )
+            Kx1 = tf.concat(
+                [
+                    ker(Z, query_points_1)[None, ...]
+                    for ker, Z in zip(kernel.kernels, inducing_points)
+                ],
+                axis=0,
+            )  # [..., L, M, A]
+            Kx2 = tf.concat(
+                [
+                    ker(Z, query_points_2)[None, ...]
+                    for ker, Z in zip(kernel.kernels, inducing_points)
+                ],
+                axis=0,
+            )  # [L, M, B]
+            K12 = tf.concat(
+                [ker(query_points_1, query_points_2)[None, ...] for ker in kernel.kernels], axis=0
+            )  # [L, M, B]
+        else:
+            K = kernel(inducing_points, full_cov=True, full_output_cov=False)  # [L, M, M]
+            Kx1 = kernel(
+                inducing_points, query_points_1, full_cov=True, full_output_cov=False
+            )  # [..., L, M, A]
+            Kx2 = kernel(
+                inducing_points, query_points_2, full_cov=True, full_output_cov=False
+            )  # [L, M, B]
+            K12 = kernel(
+                query_points_1, query_points_2, full_cov=True, full_output_cov=False
+            )  # [..., L, A, B]
+    else:
+        K = kernel(inducing_points)  # [M, M]
+        Kx1 = kernel(inducing_points, query_points_1)  # [..., M, A]
+        Kx2 = kernel(inducing_points, query_points_2)  # [M, B]
+        K12 = kernel(query_points_1, query_points_2)  # [..., A, B]
+
+    if len(tf.shape(K)) == 2:
+        K = tf.repeat(tf.expand_dims(K, -3), num_latent, axis=-3)
+        Kx1 = tf.repeat(tf.expand_dims(Kx1, -3), num_latent, axis=-3)
+        Kx2 = tf.repeat(tf.expand_dims(Kx2, -3), num_latent, axis=-3)
+        K12 = tf.repeat(tf.expand_dims(K12, -3), num_latent, axis=-3)
+    elif len(tf.shape(K)) > 3:
+        raise NotImplementedError(
+            "Covariance between points is not supported " "for kernels of type " f"{type(kernel)}."
```

Comment:
we should probably remove this now? I'm not sure sure this a case for which the shape is larger than 3. But we could throw this error when asserting shapes instead, see below